File: graphs/Questions/Shortest_Path_Maze/Python/Shortest_Path_Maze.py
def isValid(x, y, dist, mat, m, n):
    return x>=0 and x < m and y >=0 and y<n and dist[x][y] == -1 and mat[x][y] != '#'

dist = [[-1 for i in range(1001)] for i in range(1001)]
mat = [[-1 for i in range(1001)] for i in range(1001)]

m, n = map(int,input().split(" "))

ci = [-1,0,0,1]
cj = [0,-1,1,0]

for i in range(m):
    string = input()
    for j in range(n):
        mat[i][j] = string[j]
        dist[i][j] = -1
        if string[j]== 'S':
            srcx = i
            srcy = j
        elif string[j] == 'E':
            exitx = i
            exity = j
Q = []

dist[srcx][srcy] = 0
Q.append((srcx,srcy))
while len(Q):
    x = Q[0][0]
    y = Q[0][1]
    if x == exitx and y == exity:
        break
    Q.pop(0)
    for i in range(4):
        curr_x = x+ci[i]
        curr_y = y+cj[i]
        if isValid(curr_x, curr_y, dist, mat, m, n):
            Q.append((curr_x, curr_y))
            dist[curr_x][curr_y] = dist[x][y] + 1

print(dist[exitx][exity])
    

File: graphs/Questions/Shortest_Path_Maze/C++/Shortest_Path_Maze.cpp
// Question Link: https://www.codechef.com/problems/DAY4P2

#include <bits/stdc++.h>
using namespace std;
//To check if the location to traverse is valid
bool IsValid(int curr_x, int curr_y, int dist[][1001], 
	         char mat[][1001], int m,int n){
	return curr_x >= 0 && curr_x < m      //To make sure the coordinates
        && curr_y >= 0 && curr_y < n      //are not out of bounds
        && dist[curr_x][curr_y] == -1     //To check if the location has been visited or not
        && mat[curr_x][curr_y] != '#';    //To check if the location is a wall
}
int main() {
    int m, n, i, j, x, y;
    //Storing distances to a particular point
    int dist[1001][1001];
    char mat[1001][1001];
    //Set all distances to -1
    memset(dist, -1, sizeof(dist[0][0]) * 1001 * 1001);
    int srcx, srcy, exitx, exity;
    //To get the four adjacent cells
    int ci[] = {-1, 0, 0, 1}; 
    int cj[] = {0, -1, 1, 0};
    cin >> m >> n;
    for (i=0; i<m; i++) {
        string str; cin >> str;
        for (j=0; j<n; j++) {
            mat[i][j] = str[j];
            if (str[j] == 'S') {
                srcx = i; srcy = j;
            } 
            else if (str[j] == 'E') {
                exitx = i; exity = j;
            }
        }
    }
	dist[srcx][srcy] = 0;
    queue< pair<int, int> > Q;
    Q.push({srcx, srcy});
    while (!Q.empty()) {
        x = Q.front().first;
        y = Q.front().second;
        //If we have found the path to end cell
        if(x==exitx&&y==exity){
        	break;
        }
        Q.pop();
        //To check all 4 combinations
        for (int i=0; i<4; i++) {
            int curr_x = x+ci[i];
            int curr_y = y+cj[i];
            if (IsValid(curr_x, curr_y, dist, mat, m, n)) {
                //For pushing valid paths into queue and
                //Incrementing distance
                Q.push({curr_x, curr_y});
                dist[curr_x][curr_y] = dist[x][y] + 1;
            }
        }
    }

    cout << dist[exitx][exity] << endl;

    return 0;
}
